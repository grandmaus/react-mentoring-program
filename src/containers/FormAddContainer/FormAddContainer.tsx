import React, { memo } from 'react';
import FormAdd from '../../components/FormAdd/FormAdd';
import { useVisibility } from '../../hooks/useVisibility.hooks';

const FormAddContainer = () => {
  const { visibility, handleToggleVisibility } = useVisibility();

  return <FormAdd visibility={visibility} closeHandler={handleToggleVisibility} />;
};

export default memo(FormAddContainer);
