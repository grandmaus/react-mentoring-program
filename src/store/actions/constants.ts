export const FETCH_MOVIES = 'FETCH_MOVIES';
export const FETCH_SORTED_MOVIES = 'FETCH_SORTED_MOVIES';
export const FETCH_FILTERED_MOVIES = 'FETCH_FILTERED_MOVIES';
export const FETCH_SEARCHED_MOVIES = 'FETCH_SEARCHED_MOVIES';
