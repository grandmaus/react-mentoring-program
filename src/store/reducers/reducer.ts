import {
  FETCH_FILTERED_MOVIES,
  FETCH_MOVIES,
  FETCH_SORTED_MOVIES,
  FETCH_SEARCHED_MOVIES
} from '../actions/constants';
import { State } from '../types';

export const initialState: State = {
  movies: [],
  searchedMovies: [],
  genre: 'All',
  sort: 'releaseDate'
};

export const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case FETCH_MOVIES:
      return {
        ...state,
        movies: action.payload.data,
        searchedMovies: action.payload.data
      };
    case FETCH_SEARCHED_MOVIES:
      return {
        ...state,
        movies: action.payload.data,
        searchedMovies: action.payload.data
      };
    case FETCH_SORTED_MOVIES:
      return {
        ...state,
        movies: action.payload.data,
        searchedMovies: action.payload.data
      };
    case FETCH_FILTERED_MOVIES:
      return {
        ...state,
        movies: action.payload.data,
        searchedMovies: action.payload.data
      };
    default:
      return state;
  }
};
