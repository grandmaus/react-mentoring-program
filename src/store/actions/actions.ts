import { EDIT_MOVIE, FETCH_MOVIES, SET_FILTER_GENRE, SET_SEARCH_TEXT, SET_SORTING_TYPE, ADD_MOVIE } from './constants';
import { ENDPOINT } from '../constants';
import { Movie } from '../types';

export const requestMovies = (movies: ReadonlyArray<Movie>) => ({
  type: FETCH_MOVIES,
  movies,
});

export const editMovie = (editedMovies: ReadonlyArray<Movie>) => ({
  type: EDIT_MOVIE,
  editedMovies,
});

export const addMovie = (addedMovie: ReadonlyArray<Movie>) => ({
  type: ADD_MOVIE,
  addedMovie,
});

export const setSearchText = (searchText: string) => ({
  type: SET_SEARCH_TEXT,
  searchText,
});

export const setFilterGenre = (genre: string) => ({
  type: SET_FILTER_GENRE,
  genre,
});

export const setSortingType = (sortingType: string) => ({
  type: SET_SORTING_TYPE,
  sortingType,
});

export const fetchMovies = (text = '', sort = 'release_date', filter = '') => async (dispatch: any) => {
  const encodedText = encodeURIComponent(text);
  const response = await fetch(
    `${ENDPOINT}?sortBy=${sort}&sortOrder=desc&search=${encodedText}&searchBy=title&filter=${filter}&limit=35`,
  );
  const data = await response.json();

  dispatch(setSearchText(text));
  dispatch(setFilterGenre(filter));
  dispatch(setSortingType(sort));
  return dispatch(requestMovies(data));
};

export const deleteMovies = (id: number) => async (dispatch: any) => {
  await fetch(`${ENDPOINT}/${id}`, {
    method: 'DELETE',
  });
  return dispatch(fetchMovies());
};

export const fetchEditMovie = (movie: Movie) => async (dispatch: any) => {
  const response = await fetch(`${ENDPOINT}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(movie),
  });
  const data = await response.json();

  return dispatch(editMovie(data));
};

export const fetchAddMovie = (movie: Movie) => async (dispatch: any) => {
  const response = await fetch(`${ENDPOINT}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(movie),
  });
  const data = await response.json();

  return dispatch(addMovie(data));
};
