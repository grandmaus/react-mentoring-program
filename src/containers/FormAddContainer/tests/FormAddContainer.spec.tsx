import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { enableFetchMocks } from 'jest-fetch-mock';
import FormAddContainer from '../FormAddContainer';
import { initialState } from '../../../store/reducers/reducer';
import { fetchAddMovie } from '../../../store/actions/actions';

enableFetchMocks();

jest.mock('../../../store/actions/actions', () => {
  return {
    fetchAddMovie: jest.fn(() => () => ({})),
  };
});

describe('FormAddContainer', () => {
  const store = configureStore([thunk])(initialState);
  const handleToggleVisibility = jest.fn();
  const container = (
    <Provider store={store}>
      <FormAddContainer visibility handleToggleVisibility={handleToggleVisibility} />
    </Provider>
  );
  Enzyme.configure({ adapter: new Adapter() });

  it('should render correctly', () => {
    const form = renderer.create(container).toJSON();

    expect(form).toMatchSnapshot();
  });

  it('should dispatch fetchAddMovie action when form submitted', () => {
    const form = mount(container).find('Formik');
    const mockValues = {
      title: 'Title',
      release_date: '1.01.2020',
      poster_path: 'path',
      genres: ['Genre'],
      overview: 'Overview',
      runtime: 123,
    };

    form.props().onSubmit(mockValues as any);

    expect(fetchAddMovie).toHaveBeenCalledWith(mockValues);
  });
});
