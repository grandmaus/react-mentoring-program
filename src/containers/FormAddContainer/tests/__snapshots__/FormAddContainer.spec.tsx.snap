// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormAddContainer should render correctly 1`] = `
Array [
  <div
    className="sc-AxirZ igpqqj"
  >
    <button
      className="sc-AxhCb kKvuVx"
      onClick={[MockFunction]}
    />
    <p
      className="sc-AxhUy eKDZxu"
    >
      Add movie
    </p>
    <form
      className="sc-AxjAm StDqN"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="sc-Axmtr dXgRrs"
      >
        <label
          className="sc-AxheI jiCVzi"
          htmlFor="title"
        >
          Title
        </label>
        <input
          className="sc-AxgMl cKlSOK"
          id="title"
          name="title"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Title here"
          type="text"
          value=""
        />
      </div>
      <div
        className="sc-Axmtr dXgRrs"
      >
        <label
          className="sc-AxheI jiCVzi"
          htmlFor="release_date"
        >
          Release date
        </label>
        <input
          className="sc-AxgMl cKlSOK"
          id="release_date"
          name="release_date"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Select date"
          type="date"
          value=""
        />
      </div>
      <div
        className="sc-Axmtr dXgRrs"
      >
        <label
          className="sc-AxheI jiCVzi"
          htmlFor="poster_path"
        >
          Movie URL
        </label>
        <input
          className="sc-AxgMl cKlSOK"
          id="poster_path"
          name="poster_path"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Movie URL here"
          type="text"
          value=""
        />
      </div>
      <div
        className="sc-fznZeY kaBQlm"
      >
        <label
          className="sc-fzplWN jwnwOR"
        >
          Genre
        </label>
        <button
          className="sc-fzpans bpfiuc"
          onClick={[Function]}
          type="button"
        >
          Select genre
        </button>
        <ul
          className="sc-fznyAO fEuoJQ"
        >
          <li
            className="sc-fznKkj ebpiHJ"
          >
            <input
              id="Drama"
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="Drama"
            />
            <label
              className="sc-fzokOt hEcmYN"
              htmlFor="Drama"
            >
              Drama
            </label>
          </li>
          <li
            className="sc-fznKkj ebpiHJ"
          >
            <input
              id="Romance"
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="Romance"
            />
            <label
              className="sc-fzokOt hEcmYN"
              htmlFor="Romance"
            >
              Romance
            </label>
          </li>
          <li
            className="sc-fznKkj ebpiHJ"
          >
            <input
              id="Horror"
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="Horror"
            />
            <label
              className="sc-fzokOt hEcmYN"
              htmlFor="Horror"
            >
              Horror
            </label>
          </li>
          <li
            className="sc-fznKkj ebpiHJ"
          >
            <input
              id="Fantasy"
              name="genres"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value="Fantasy"
            />
            <label
              className="sc-fzokOt hEcmYN"
              htmlFor="Fantasy"
            >
              Fantasy
            </label>
          </li>
        </ul>
      </div>
      <div
        className="sc-Axmtr dXgRrs"
      >
        <label
          className="sc-AxheI jiCVzi"
          htmlFor="overview"
        >
          Overview
        </label>
        <input
          className="sc-AxgMl cKlSOK"
          id="overview"
          name="overview"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Overview here"
          type="text"
          value=""
        />
      </div>
      <div
        className="sc-Axmtr dXgRrs"
      >
        <label
          className="sc-AxheI jiCVzi"
          htmlFor="runtime"
        >
          Runtime
        </label>
        <input
          className="sc-AxgMl cKlSOK"
          id="runtime"
          name="runtime"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Runtime here"
          type="number"
          value={0}
        />
      </div>
      <div
        className="sc-fzozJi eWnHqs"
      >
        <button
          className="sc-fzoLsD dHOvdw"
          color="#f65251"
          type="reset"
          width="200px"
        >
          Reset
        </button>
        <button
          className="sc-fzoLsD bjrBUL"
          color="#ffffff"
          type="submit"
          width="200px"
        >
          Submit
        </button>
      </div>
    </form>
  </div>,
  <div
    className="sc-AxiKw eqBQW"
  />,
]
`;
